libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,579 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Stack layout generator for Yul to EVM code generation.
+ */
+
+#include <libyul/backends/evm/StackLayoutGenerator.h>
+
+#include <libyul/backends/evm/StackHelpers.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/cxx20.h>
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/concat.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace std;
+
+StackLayout StackLayoutGenerator::run(CFG const& _cfg)
+{
+	StackLayout stackLayout;
+	StackLayoutGenerator{stackLayout, {}}.processEntryPoint(*_cfg.entry);
+
+	for (auto& functionInfo: _cfg.functionInfo | ranges::views::values)
+		StackLayoutGenerator{stackLayout, functionInfo.returnVariables}.processEntryPoint(*functionInfo.entry);
+
+	return stackLayout;
+}
+
+StackLayoutGenerator::StackLayoutGenerator(StackLayout& _layout, vector<VariableSlot> _currentFunctionReturnVariables):
+	m_layout(_layout),
+	m_currentFunctionReturnVariables(move(_currentFunctionReturnVariables))
+{
+}
+
+namespace
+{
+
+/// @returns the ideal stack to have before executing an operation that outputs @a _operationOutput, s.t.
+/// shuffling to @a _post is cheap.
```

Comment:
How can I know the stack layout before the operation if I don't know what the operation consumes?

Reply:
Ok, it's the layout *except* the inputs of the operation itself...

Reply:
ok, right, because the inputs have to be at the top and if they need to be dupped, this is another job.

Reply:
Yeah - the final pre-operation stack will be the result of this monster here plus the operation inputs.
An input might occur among the ``PreviousSlot`` things as well, that basically means the input will be dupped, or if it *doesn't* occur among the PreviousSlot's, then that basically means it will be swapped (i.e. the only occurrence will be in the operation input-part and will be consumed by the operation).